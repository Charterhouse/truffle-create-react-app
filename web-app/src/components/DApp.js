import React from 'react'
import { Link } from 'react-router-dom'
import withWeb3 from 'web3/withWeb3'

// Demonstration of a basic dapp with the withWeb3 higher-order component
class DApp extends React.Component {
  state = { balance: null }

  async componentDidMount () {
    await this.getValue()
  }

  storeValue = async () => {
    const { accounts, contract } = this.props
    await this.getValue()
    await contract.set(this.state.balance + 5, { from: accounts[0] })
    await this.getValue()
  }

  getValue = async () => {
    const { accounts, contract } = this.props
    const response = await contract.get.call({ from: accounts[0] })
    this.setState({ balance: response.toNumber() })
  }

  render () {
    // Uncomment to use web3, accounts or the contract:
    // const { web3, accounts, contract } = this.props
    const { balance } = this.state
    return (
      <div style={{ padding: '20px' }}>
        <h1>My DApp</h1>
        <div>
          <p style={{ display: 'inline-block', marginBottom: '20px' }}>Current Balance: {balance}</p>
          <button style={{ marginLeft: '20px' }} onClick={this.getValue}>Refresh...</button>
        </div>
        <button onClick={this.storeValue}>Add 5 to the account balance</button>
        <div style={{ marginTop: '20px' }}>
          <Link to='/'>Home</Link>
          <Link to='/accounts' style={{ marginLeft: '20px' }}>My Accounts</Link>
        </div>
      </div>
    )
  }
}

export default withWeb3(DApp)
