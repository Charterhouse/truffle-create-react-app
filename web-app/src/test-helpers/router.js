import React from 'react'
import { MemoryRouter as Router, Route } from 'react-router-dom'

export const inRouter = (Component, path) => {
  return (
    <Router initialEntries={[ path ]} initialIndex={0}>
      <Route exact path={path} component={Component} />
    </Router>
  )
}

export const renderInRouter = (componentRenderFunction, path) => {
  return (
    <Router initialEntries={[ path ]} initialIndex={0}>
      <Route exact path={path} render={componentRenderFunction} />
    </Router>
  )
}
