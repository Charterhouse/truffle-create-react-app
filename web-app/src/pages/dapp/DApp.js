import React from 'react'
import { Wrapper } from 'components/wrapper'
import { AppNavigation } from 'components/navigation'

// Demonstration of a basic dapp with the withWeb3 higher-order component
class DApp extends React.Component {
  state = { balance: null }

  async componentDidMount () {
    await this.getValue()
  }

  storeValue = async () => {
    const { accounts, contract } = this.props
    await this.getValue()
    await contract.set(this.state.balance + 5, { from: accounts[0] })
    await this.getValue()
  }

  getValue = async () => {
    const { accounts, contract } = this.props
    const response = await contract.get.call({ from: accounts[0] })
    this.setState({ balance: response.toNumber() })
  }

  render () {
    // Uncomment to use web3, accounts or the contract:
    // const { web3, accounts, contract } = this.props
    const { balance } = this.state
    return (
      <Wrapper>
        <h1>My DApp</h1>
        <div>
          <P>Current Balance: {balance}</P>
          <Button leftMargin onClick={this.getValue}>Refresh...</Button>
        </div>
        <Button onClick={this.storeValue}>Add 5 to the account balance</Button>
        <AppNavigation location={this.props.location} />
      </Wrapper>
    )
  }
}

const P = ({ children }) =>
  <p style={{ display: 'inline-block', marginBottom: '20px' }}>{ children }</p>

const Button = ({ children, leftMargin, ...rest }) => (
  leftMargin
    ? <button style={{ marginLeft: '20px' }} {...rest}>{ children }</button>
    : <button {...rest}>{ children }</button>
)

export { DApp }
