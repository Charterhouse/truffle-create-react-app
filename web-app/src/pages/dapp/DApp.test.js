import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { renderInRouter } from 'test-helpers/router'
import { DApp } from './DApp'

const testAccounts = [
  '0x627306090abab3a6e1400e9345bc60c78a8bef57'
]

const testContract = {
  set: (balance, obj) => {
    return new Promise(resolve => {
      resolve()
    })
  },
  get: {
    call: obj => {
      return new Promise(resolve => {
        resolve({ toNumber: () => 15 })
      })
    }
  }
}

const renderDAppInRoute = routeProps =>
  <DApp accounts={testAccounts}
    contract={testContract}
    {...routeProps} />

let dappInRoute

beforeEach(() => {
  dappInRoute = renderInRouter(
    renderDAppInRoute,
    '/dapp')
})

it('renders without crashing', async () => {
  const div = document.createElement('div')
  await ReactDOM.render(dappInRoute, div)
})

it('renders correctly', async () => {
  const dapp = await renderer
    .create(dappInRoute)
  await waitForAsyncCallsToComplete()
  expect(dapp.toJSON()).toMatchSnapshot()
})

const waitForAsyncCallsToComplete = async () => {
  // queue another item on the event loop callback queue
  // and wait for it to be executed
  await Promise.resolve()
}
