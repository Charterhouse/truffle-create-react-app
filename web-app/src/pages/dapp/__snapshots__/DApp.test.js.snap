// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "20px",
    }
  }
>
  <h1>
    My DApp
  </h1>
  <div>
    <p
      style={
        Object {
          "display": "inline-block",
          "marginBottom": "20px",
        }
      }
    >
      Current Balance: 
      15
    </p>
    <button
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "20px",
        }
      }
    >
      Refresh...
    </button>
  </div>
  <button
    onClick={[Function]}
  >
    Add 5 to the account balance
  </button>
  <div
    style={
      Object {
        "marginTop": "20px",
      }
    }
  >
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <a
      href="/accounts"
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "20px",
        }
      }
    >
      My Accounts
    </a>
  </div>
</div>
`;
